Yosemite/Yosemite/Stores/ProductAttributeTermStore.swift

```diff
@@ -0,0 +1,175 @@
+
+    private lazy var sharedDerivedStorage: StorageType = {
+        return storageManager.newDerivedStorage()
+    }()
+
+    public override init(dispatcher: Dispatcher, storageManager: StorageManagerType, network: Network) {
+        self.remote = ProductAttributeTermRemote(network: network)
+        super.init(dispatcher: dispatcher, storageManager: storageManager, network: network)
+    }
+
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: ProductAttributeTermAction.self)
+    }
+
+    override public func onAction(_ action: Action) {
+        guard let action = action as? ProductAttributeTermAction else {
+            assertionFailure("ProductAttributeTermStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case let .synchronizeProductAttributeTerms(siteID, attributeID, onCompletion):
+            synchronizeAllProductAttributeTerms(siteID: siteID, attributeID: attributeID, fromPageNumber: 0) { error in
+                let result: Result<Void, ProductAttributeTermActionError> = {
+                    if let error = error {
+                        return .failure(error)
+                    }
+                    return .success(())
+                }()
+                onCompletion(result)
+            }
+        }
+    }
+}
+
+// MARK: - Services
+private extension ProductAttributeTermStore {
+
+    /// Synchronizes all product attribute terms associated with a given `SiteID` and `attributeID`, starting at a specific page number.
+    /// Having the correct parent `ProductAttribute` is a requirement, otherwise nothing will be saved.
+    ///
+    func synchronizeAllProductAttributeTerms(siteID: Int64,
+                                             attributeID: Int64,
+                                             synchronizedTerms: [Yosemite.ProductAttributeTerm] = [],
+                                             fromPageNumber: Int,
+                                             onCompletion: @escaping (ProductAttributeTermActionError?) -> Void) {
+        // Start fetching the initial page
+        synchronizeProductAttributeTerms(siteID: siteID,
+                                         attributeID: attributeID,
+                                         pageNumber: fromPageNumber,
+                                         pageSize: Constants.defaultMaxPageSize) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+
+            // If terms is empty, end the recursion and call `onCompletion`
+            case let .success(terms) where terms.isEmpty:
+                self.deleteStaleTerms(siteID: siteID, attributeID: attributeID, activeTerms: synchronizedTerms)
+                onCompletion(nil)
+
+            // If there could be more(non-empty) terms, request the next page recursively.
+            case let .success(terms):
```

Comment:
nit: a potential minor optimization here is to call completion if the term count is smaller than the page size (the last page is most often partially filled) so we don't have to make another API request for an empty response. If `terms.count < pageSize` is true, we can call `self.deleteStaleTerms(siteID: siteID, attributeID: attributeID, activeTerms: synchronizedTerms)` and `onCompletion` like in the case above

Reply:
Great Suggestion!